fixes: #120: Need driver update for Neo4j 4.0

conker84 opened the pull request:
List of tasks:
- [x] updated driver to Neo4j v 4.0
- [x] added support to multi-tenancy via option `spark.neo4j.database`
- [x] improved tests with Testcontainers

krisgeus commented on pom.xml:
```diff
@@ -39,15 +39,27 @@
-        <spark.version>2.4.0</spark.version>
+        <neo4j.version>4.0.0</neo4j.version>
+        <driver.version>4.0.0</driver.version>
+        <spark.version>2.4.1</spark.version>
```
Spark 2.4.4 is the latest and greatest spark version out there. Might be good to use that one.

krisgeus replied:
Other than that LGTM

moxious commented on src/main/scala/org/neo4j/spark/rdd/Neo4jRowRDD.scala:
```diff
@@ -20,31 +20,31 @@ class Neo4jRowRDD(@transient sc: SparkContext, val query: String, val parameters
+        }.iterator
+      }
+      session.readTransaction(runner)
     } finally {
```
I think this needs experimentation.   I could be wrong about this so we don't necessarily need a change but when I was playing with my own private branch adapting spark for 4.0, these "finally close the session" blocks were a big problem.   This needs more testing.

Reason is that under the new reactive driver setup, the semantics of how/when records are consumed has changed.   Spark itself is so lazy, when you exit this block the very first record hasn't been pulled yet, but you're closing the session.   In my experimentation this broke badly because basically you'd always fail to get data back from neo4j saying that the driver or the session was closed.

For unit tests on this you might need some delay, and record sets with > 1 coming back.   Or you can create a test build, play with it in zeppelin

I'll not put the same comment on every finally block, but this pattern (run the query, return, close the session/driver in a finally block) is common in the code base.

moxious replied:
See "important note on semantics" in the new java driver API docs -- I think this interacting with spark's laziness is specifically problematic.  https://neo4j.com/docs/api/java-driver/4.0/org/neo4j/driver/QueryRunner.html

conker84 replied:
@moxious thanks a lot for the feedback. We definitely need to improve this, I have in mind to do it with a new release that supports Spark Datasource APIs.

About transaction management, looking at the code I invoke `Result#list` which store the whole result for later use
https://neo4j.com/docs/api/java-driver/4.0/org/neo4j/driver/Result.html#list--

This is the only way that I found to interact with the db that allows us to use transaction functions (that should allow us to work properly with Aura)

Wdyt?

jexp replied:
it's definitely a tricky area. I think we should discuss with the drivers team what's the best interaction pattern that gives us all

- lazyness / reactive
- transaction functions
- ensure closing driver / session to avoid leaks
- support sparks lazy evaluation
- avoid pre-materialization of results

conker84 replied:
can we organize a chat with them?

moxious replied:
I'll ask if we can put together that chat.  I don't have an opinion on a solution approach because this stuff is too new to me and I don't understand the consequences either way.

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -35,79 +35,50 @@ object Executor {
+        var i = 0
+        while (i < keyCount) {
+          val value = record.get(i).asObject() match {
+            case it: util.Map[_, _] => it.asScala.toString()
```
why toString?

conker84 replied:
restored, I realized that the test `runCypherQueryWithSchemaAndMap` was wrong, I fixed it

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -35,79 +35,50 @@ object Executor {
-              row.update(i, value)
-              i = i + 1
-            }
-            if (!result.hasNext) {
```
how do we handle closing of resources now?

jexp replied:
ping?

moxious replied:
@jexp @conker84 Nigel on the drivers team said that the API still provides for the old style sessions which are not lazy/reactive, and we should use those

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -35,79 +35,50 @@ object Executor {
         }
+        row
+      })
+      new CypherResult(schema, new LazySessionIterator(it, driver, session))
```
esp. how does it work to return/leak the driver+session+result out of the spark job processing?

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -35,79 +35,50 @@ object Executor {
+      new CypherResult(schema, new LazySessionIterator(it, driver, session))
+    } catch {
+      case e: Exception => {
+        if (session.isOpen) {
```
null checks?

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -35,79 +35,50 @@ object Executor {
-
     val driver: Driver = config.driver()
-    val session = driver.session()
+    val session = driver.session(config.sessionConfig())
```
what closes the session and driver? shouldn't be there a finally or such?

jexp replied:
ping?

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -35,79 +35,50 @@ object Executor {
+        throw e
       }
-
-      if (write)
```
the read and write transactions are very important for routing in a causal cluster so we must not lose that functionality

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -35,79 +35,50 @@ object Executor {
-            result.consume()
-            close(driver, session)
-            return res
+      val result: Result = session.run(query, toJava(parameters))
```
session.run doesn't cut it, see below, we need to use tx-functions for routing and retries

jexp commented on src/main/scala/org/neo4j/spark/LazySessionIterator.scala:
```diff
@@ -0,0 +1,32 @@
+import org.neo4j.driver.Driver
+import org.neo4j.driver.Session
+
+class LazySessionIterator[T](iterator: Iterator[T], driver: Driver, session: Session) extends Iterator[T] {
```
I'm not suer this really works ... can this leak out of scope of the worker job so that the driver/session might not be closed (if no one calls close here) there is no guarantee or?

jexp commented on src/main/scala/org/neo4j/spark/LazySessionIterator.scala:
```diff
@@ -0,0 +1,32 @@
+  private lazy val it = iterator
+
+  override def hasNext: Boolean = {
+    val hasNext = it.hasNext
```
also has to close if if it.next throws an error

jexp commented on src/main/scala/org/neo4j/spark/LazySessionIterator.scala:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  private def close = {
+    if (session.isOpen) session.close()
```
null check on session and wrap with try-catch in case session.close throws  an exception

jexp commented on src/main/scala/org/neo4j/spark/Neo4jConfig.scala:
```diff
@@ -1,34 +1,41 @@
  * @since 02.03.16
  */
-case class Neo4jConfig(val url: String, val user: String = "neo4j", val password: Option[String] = None, encryptionStatus:Boolean) {
+case class Neo4jConfig(val url: String,
```
default value for encryption (not sure where "status" comes from there)

jexp commented on src/main/scala/org/neo4j/spark/Neo4jConfig.scala:
```diff
@@ -1,34 +1,41 @@
  * @author mh
  * @since 02.03.16
  */
-case class Neo4jConfig(val url: String, val user: String = "neo4j", val password: Option[String] = None, encryptionStatus:Boolean) {
```
should we also add support for multi-db here? specify the db to connect to?

conker84 replied:
There is the multi support db via config property `spark.neo4j.database`.

jexp commented on src/main/scala/org/neo4j/spark/Neo4jConfig.scala:
```diff
@@ -1,34 +1,41 @@
+    val url = sparkConf.get(s"$prefix.bolt.url", "bolt://localhost")
+    val user = sparkConf.get(s"$prefix.bolt.user", "neo4j")
+    val password: Option[String] = sparkConf.getOption(s"$prefix.bolt.password")
+    val database: Option[String] = sparkConf.getOption(s"$prefix.database")
```
why no bolt here ? seems inconsistent?

conker84 replied:
Because it's not bolt strictly related to I used `spark.neo4j.database`
Do you prefer `spark.neo4j.bolt.database`?

jexp replied:
either we use bolt everywhere or we drop it everywhere (then we have to update docs + readme)
user and password are also not related to bolt :)

moxious replied:
Driver changes in 4.0 have sadly made this one complicated.   bolt will continue to work, but it's deprecated, and implies the wrong thing for 3.5 cluster.   Neo4j is new hotness -- but *does not* work with 3.5.

My vote would be to make it `spark.neo4j.database` and use "neo4j" throughout to be supportive of the overall messaging for Neo4j.  But 100% we're going to need some special instructions for migration of the old driver, and an explanation of all of this bolt, bolt+routing, and neo4j craziness.

conker84 replied:
@jexp you're right I'm removing bolt from everywhere and updating the docs.

jexp commented on src/main/scala/org/neo4j/spark/Neo4jConfig.scala:
```diff
@@ -1,34 +1,41 @@
+    val user = sparkConf.get(s"$prefix.bolt.user", "neo4j")
+    val password: Option[String] = sparkConf.getOption(s"$prefix.bolt.password")
+    val database: Option[String] = sparkConf.getOption(s"$prefix.database")
+    val encryptionStatus : Boolean = sparkConf.getBoolean(s"$prefix.bolt.encryption.status", defaultValue = false)
```
no need for ".status"

conker84 replied:
done

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -206,38 +145,35 @@ object Neo4jDataFrame {
       val driver = config.driver()
-      val session = driver.session()
+      val session = driver.session(config.sessionConfig())
+      val result = session.run(query, params)
```
sorry, doesn't work, we need the transaction function for routing and retries

jexp commented on src/main/scala/org/neo4j/spark/rdd/Neo4jRowRDD.scala:
```diff
@@ -20,12 +20,11 @@ class Neo4jRowRDD(@transient sc: SparkContext, val query: String, val parameters
-      val result: StatementResult = session.run(query, parameters.toMap.mapValues(_.asInstanceOf[AnyRef]).asJava)
-
-      result.asScala.map(record => {
+      val result = session.run(query, parameters.toMap.mapValues(_.asInstanceOf[AnyRef]).asJava)
```
same here

jexp commented on src/main/scala/org/neo4j/spark/rdd/Neo4jRowRDD.scala:
```diff
@@ -39,15 +38,17 @@ class Neo4jRowRDD(@transient sc: SparkContext, val query: String, val parameters
+      new LazySessionIterator(it, driver, session)
+    } catch {
+      case e: Exception => {
+        if (session.isOpen) {
```
null checks and try-catch for session closing 
that's why I wrapped this in a function that I could reuse in all the places !

conker84 replied:
ok

jexp commented on src/main/scala/org/neo4j/spark/rdd/Neo4jTupleRDD.scala:
```diff
@@ -14,21 +19,21 @@ class Neo4jTupleRDD(@transient sc: SparkContext, val query: String, val paramete
-          driver.close()
-        }
-        res
+      val result: Result = session.run(query, parameters.toMap.asJava)
```
same here with tx function

jexp commented on src/main/scala/org/neo4j/spark/rdd/Neo4jTupleRDD.scala:
```diff
@@ -14,21 +19,21 @@ class Neo4jTupleRDD(@transient sc: SparkContext, val query: String, val paramete
+      new LazySessionIterator(it, driver, session)
+    } catch {
+      case e: Exception => {
+        if (session.isOpen) {
```
same here

jexp commented on src/test/scala/org/neo4j/spark/SparkConnectorScalaSuiteIT.scala:
```diff
@@ -0,0 +1,48 @@
+
+
+object SparkConnectorScalaSuiteIT {
+  val server: Neo4jContainerExtension = new Neo4jContainerExtension("neo4j:4.0.0-enterprise")
```
should we use latest-enterprise?

conker84 replied:
done

jexp commented on src/test/scala/org/neo4j/spark/SparkConnectorScalaSuiteIT.scala:
```diff
@@ -0,0 +1,48 @@
+        case _ => //
+      }
+      Assume.assumeTrue("Neo4j container is not started", server.isRunning)
+      conf = new SparkConf().setAppName("neoTest")
```
Should we set up spark with testcontainers too?

conker84 replied:
Yes absolutely, can we do it in a separate PR?

jexp replied:
sure

jexp commented on README.md:
```diff
@@ -69,7 +69,7 @@ Otherwise set the `spark.neo4j.bolt.url` in your `SparkConf` pointing e.g. to `b
-Otherwise, you can either ignore `spark.neo4j.bolt.encryption.status` or  set  `spark.neo4j.bolt.encryption.status` to `false` in your `SparkConf`.
+Otherwise, you can either ignore `spark.neo4j.bolt.encryption` or  set  `spark.neo4j.bolt.encryption` to `false` in your `SparkConf`.
 
 ## Builder API
```
can you make sure in the readme the other version numbers that are mentioned are fixed / correct?

conker84 replied:
Done

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -21,7 +22,9 @@ object Neo4jDataFrame {
+                    renamedColumns: Map[String, String] = Map.empty,
+                    partitions: Int = 1,
+                    unwindBatchSize: Int = 10000): Unit = {
     dataFrame.cache()
```
should this by default? isn't that something that the user wants to control?

conker84 replied:
ok

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -32,56 +35,66 @@ object Neo4jDataFrame {
         MERGE (target:${target._1.quote} {${targetLabel.quote} : row.target.${targetLabel.quote}}) ON CREATE SET target += row.target
```
we should split this up, into two statements like in saveGraph

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -32,56 +35,66 @@ object Neo4jDataFrame {
         UNWIND $$rows as row
-        CREATE (node:${nodes._1.quote} {${nodeLabel.quote} : row.source.${nodeLabel.quote}})
-        SET node = row.node_properties
+        CREATE (node:${nodes._1.quote} {${nodeLabel.quote} : row.node_properties.${nodeLabel.quote}})
```
possibly make this configurable via parameter (GH issue)

jexp replied:
MERGE vs. CREATE I mean

conker84 replied:
I added it in this PR

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -32,56 +35,66 @@ object Neo4jDataFrame {
-    } finally {
-      if (session.isOpen) session.close()
-      driver.close()
+  def execute(session: Session, query: String, parameters: java.util.Map[String, AnyRef], write: Boolean = false): ResultSummary = {
```
should this be private?

conker84 replied:
done

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -111,88 +124,26 @@ object Neo4jDataFrame {
-//  }
-
   def apply(sqlContext: SQLContext, query: String, parameters: java.util.Map[String, AnyRef]): DataFrame = {
     val limitedQuery = s"$query LIMIT 1"
```
don't remember where this was from ...

jexp replied:
a just to determine the datatypes

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -206,13 +157,13 @@ object Neo4jDataFrame {
             val result = tx.run(query, params)
             if (!result.hasNext) throw new RuntimeException("Can't determine schema from empty result")
-            result.asScala.map(record => {
+            result.list().asScala.map(record => {
```
this is potentially a lot that's eagerly pulled ... (think 10 or 100M or 1bn rows ...) ?
I meant to defer the materialization as the iterator is used by spark
and close resources after the last record has been consumed.

jexp replied:
ping?

jexp commented on src/test/scala/org/neo4j/spark/SparkConnectorScalaBaseTSE.scala:
```diff
@@ -0,0 +1,49 @@
+  @Before
+  @throws[Exception]
+  def start() {
+    session().run("MATCH (n) DETACH DELETE n").consume()
```
for large tests this might exceed memory

conker84 replied:
Yes I agree with you, I did that in order to start the Neo4j container just once, for large tests  we should think about a different strategy.

jexp commented on src/test/scala/org/neo4j/spark/SparkConnectorScalaBaseTSE.scala:
```diff
@@ -0,0 +1,49 @@
+  private var driver: Driver = _
+
+  def session(): Session = {
+    if (driver == null) {
```
shouldn't we allocate teh driver at the beginning of the test?

conker84 replied:
The container instance is one for all tests so I think we can do it here

jexp commented on src/main/scala/org/neo4j/spark/utils/Neo4jUtils.scala:
```diff
@@ -0,0 +1,16 @@
+object Neo4jUtils {
+
+  def close(driver: Driver, session: Session): Unit = {
+    if (session != null && session.isOpen) {
```
wrap with try-catch to guarantee to close the driver even if closing the session fails for some reason,
want to avoid leaks

conker84 replied:
perfect, done

jexp commented on src/main/scala/org/neo4j/spark/rdd/Neo4jTupleRDD.scala:
```diff
@@ -14,21 +15,19 @@ class Neo4jTupleRDD(@transient sc: SparkContext, val query: String, val paramete
+      val txWork = new TransactionWork[Iterator[Seq[(String, AnyRef)]]] {
+        override def execute(tx: Transaction): Iterator[Seq[(String, AnyRef)]] = {
+          val result: Result = tx.run(query, parameters.toMap.asJava)
+          result.list().asScala.map((record) => {
```
same as in the other place, i'm a bit afraid of materializing all results eagerly and not as spark consumes the iterator ...

jexp replied:
ping @conker84 ?

jexp commented on README.md:
```diff
@@ -255,22 +255,25 @@ There are a few different RDD's all named `Neo4jXxxRDD`
+    * optional `renamedColumns` parameter - can be used to create a relationship between nodes having the same properties.
+    * the `partitions` parameter defines the number of partitions of the dataframe will be repartitioned into. It defines the level of the parallelism, so consider that it could lead to transaction locks if the same node gets managed concurrently;
+    * the `unwindBatchSize` parameter defines for each partition the batch size sent to Neo4j.<br/>
  *For example*: `(keanu:Person {name: 'Keanu'})-[:ACTED_WITH]->(Laurence:Person {name: "Laurence"})` requires a DataFrame with 2 `name` columns which is not possible.
```
the example is for renamedColums, need to keep it up there.

conker84 replied:
done

jexp commented on src/main/scala/org/neo4j/spark/rdd/Neo4jRowRDD.scala:
```diff
@@ -20,34 +21,32 @@ class Neo4jRowRDD(@transient sc: SparkContext, val query: String, val parameters
+      val txWork = new TransactionWork[Iterator[Row]] {
+        override def execute(tx: Transaction): Iterator[Row] = {
+          val result = tx.run(query, parameters.toMap.mapValues(_.asInstanceOf[AnyRef]).asJava)
+          result.list().asScala.map(record => {
```
same here ...

jexp commented on src/main/scala/org/neo4j/spark/rdd/Neo4jRowRDD.scala:
```diff
@@ -20,34 +21,32 @@ class Neo4jRowRDD(@transient sc: SparkContext, val query: String, val parameters
+            val res = if (keyCount == 0) Row.empty
+            else if (keyCount == 1) Row(convert(record.get(0).asObject()))
+            else {
+              val builder = Seq.newBuilder[AnyRef]
```
seed with keyCount

conker84 replied:
done

jexp commented on src/test/scala/org/neo4j/spark/Neo4jGraphScalaTSE.scala:
```diff
@@ -0,0 +1,86 @@
+
+  @Before
+  @throws[Exception]
+  def setUp {
```
should we do these tests with larger numbers that exceed partition sizes?

jexp replied:
ping @conker84

conker84 replied:
Yes I'll do that in a separate issue ok? I created #123

conker84 commented:
> Some of my questions din't have a feedback can you have another look.
> Could you summarize the changes you did after my last feedback esp. around the tx functions result.list resource (close) handling etc?

I answered to the question as I found the solution, I'm still working on the others (tx-fun and resource handling) I'll be back once I found a solution

jexp commented:
@conker84 did you test the changes with some larger (e.g. 10M nodes) read and write setups? (with multiple partitions)
And with the aura instance too?

conker84 commented:
@jexp as we discussed on Slack I implemented the Unmanaged Transaction flow.
I added a test `after` each method that checks if there are leaked transaction in that test execution.
I get the current active sessions at the container startup via the following code:

https://github.com/neo4j-contrib/neo4j-spark-connector/blob/1e677e7d8266716abe0fc33b5318f29f080f955a/src/test/scala/org/neo4j/spark/SparkConnectorScalaSuiteIT.scala#L66-L76

And here:

https://github.com/neo4j-contrib/neo4j-spark-connector/blob/1e677e7d8266716abe0fc33b5318f29f080f955a/src/test/scala/org/neo4j/spark/SparkConnectorScalaSuiteIT.scala#L45

Then in the after phase of each test I do the following check:

https://github.com/neo4j-contrib/neo4j-spark-connector/blob/1e677e7d8266716abe0fc33b5318f29f080f955a/src/test/scala/org/neo4j/spark/SparkConnectorScalaBaseTSE.scala#L48-L53

I tried to apply al your feedback if I lost a comment please apologise in advance.

I also tried with cloud-to-cloud import from Databricks CE to AURA (the smallest instance),
I was able to import a csv of 2 milion rows (the max size without getting an OOM from Databricks CE) in around 44 mins. Please check the following image:

<img width="1594" alt="Schermata 2020-03-30 alle 22 52 37" src="https://user-images.githubusercontent.com/1833335/78018923-0e6f3080-734f-11ea-9795-346ca069f930.png">

jexp commented on README.md:
```diff
@@ -62,17 +62,18 @@ In your pom.xml, add:
 
 ## Config
 
-If you're running Neo4j on localhost with the default ports, you only have to configure your password in `spark.neo4j.bolt.password=<password>`.
```
can we highlight prominently that the config has changed?
do we still support the old config as a fallback?

conker84 replied:
Should we support the old config?

conker84 replied:
I added this: `**Nb.**  The old spark config prefix `spark.neo4j.bolt` is deprecated and will be removed in the next release.`

jexp commented on pom.xml:
```diff
@@ -39,15 +39,28 @@
-        <driver.version>1.7.5</driver.version>
-        <spark.version>2.4.0</spark.version>
+        <neo4j.version>4.0.1</neo4j.version>
+        <driver.version>4.0.1</driver.version>
```
4.0.2 ?

conker84 replied:
ok

conker84 replied:
The last one on maven central is:
https://mvnrepository.com/artifact/org.neo4j.driver/neo4j-java-driver

jexp replied:
oh sorry

jexp commented on pom.xml:
```diff
@@ -39,15 +39,28 @@
-        <neo4j.version>3.5.12</neo4j.version>
-        <driver.version>1.7.5</driver.version>
-        <spark.version>2.4.0</spark.version>
+        <neo4j.version>4.0.1</neo4j.version>
```
4.0.3

conker84 replied:
ok

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -1,32 +1,19 @@
 
 object Executor {
 
-  def convert(value: AnyRef): Any = value match {
```
what happend with this?

conker84 replied:
I'm sorry something went wrong when I rebased the code.
I'll re-add this and others part

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -48,75 +37,39 @@ object Executor {
+      val row = new Array[Any](keyCount)
+      var i = 0
+      while (i < keyCount) {
+        val value = record.get(i).asObject() match {
```
datetime conversion was removed!! please re-add

conker84 replied:
I'm sorry something went wrong when I rebased the code.
I'll re-add this and others part

jexp commented on src/main/scala/org/neo4j/spark/Executor.scala:
```diff
@@ -48,75 +37,39 @@ object Executor {
-            val row = new Array[Any](keyCount)
-            var i = 0
-            while (i < keyCount) {
-              val value = convert(record.get(i).asObject())
```
pleas re-add this bit

conker84 replied:
I'm sorry something went wrong when I rebased the code.
I'll re-add this and others part

jexp commented on src/main/scala/org/neo4j/spark/Neo4jConfig.scala:
```diff
@@ -1,34 +1,44 @@
 }
 
 object Neo4jConfig {
-  val prefix = "spark.neo4j.bolt."
```
should we as a fallback support the old syntax for a version or two?

conker84 replied:
Oki

conker84 commented:
@jexp I updated the PR thanks a lot for the feedback!

conker84 commented:
@jexp I'm investigating why is falling on Travis while in my local env the build is successful

conker84 commented:
@jexp it's green now

conker84 commented:
I wrapped the connection check with and `assertEventually`:

https://github.com/neo4j-contrib/neo4j-spark-connector/blob/37a69a39faa5fe972ef86cea5514167231eaf717/src/test/scala/org/neo4j/spark/SparkConnectorScalaBaseTSE.scala#L55-L61

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -24,67 +23,78 @@ object Neo4jDataFrame {
+      override def execute(tx: Transaction): ResultSummary =
+        tx.run(query, parameters).consume()
+    }
+    if (write) {
```
do we still need all this incl. the TransactionWork if we have the defaultTxMode?

conker84 replied:
Just because I want to left the old implementation as much as possible in order to reduce the impact of the new changes.
If you want I can add the same strategy here and in the other parts

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -265,7 +199,7 @@ object CypherTypes {
     case "BOOL" => BOOLEAN
     case "DATETIME" => DATETIME
-	case "TIME" => TIME
+    case "TIME" => DATETIME
```
Are you sure?

jexp replied:
TIME -> DATETIME?

conker84 replied:
Just because otherwise this will not work:

https://github.com/neo4j-contrib/neo4j-spark-connector/blob/70ae1f647b6f66eddee666e55d048faded17f5f0/src/main/scala/org/neo4j/spark/Executor.scala#L24

jexp commented on src/main/scala/org/neo4j/spark/rdd/Neo4jTupleRDD.scala:
```diff
@@ -13,23 +16,9 @@ class Neo4jTupleRDD(@transient sc: SparkContext, val query: String, val paramete
-      driver.close()
-    }
+    new Neo4jSessionAwareIterator(config, query, parameters.toMap.asJava, false).map(record => {
+      record.asMap().asScala.toSeq
```
do we have to convert to Spark types here? Perhaps separate PR/issue

conker84 replied:
Done in this one

jexp commented on src/main/scala/org/neo4j/spark/utils/Neo4jSessionAwareIterator.scala:
```diff
@@ -0,0 +1,62 @@
+      }
+      hasNext
+    } catch {
+      case _ => {
```
Do we plan to report errors somewhere?

conker84 replied:
Added a logger

jexp commented on src/main/scala/org/neo4j/spark/utils/Neo4jSessionAwareIterator.scala:
```diff
@@ -0,0 +1,62 @@
+      if (result != null) {
+        result.consume()
+      }
+    } finally {
```
should we catch exceptions here? as they are probably not relevant.

conker84 replied:
What do you mean?

jexp commented on src/main/scala/org/neo4j/spark/utils/Neo4jUtils.scala:
```diff
@@ -0,0 +1,58 @@
+      classOf[SessionExpiredException], classOf[ServiceUnavailableException] // retry on the same exceptions the driver does [1]
+    )
+    .retryOnException(new function.Predicate[Throwable] {
+      override def test(exception: Throwable): Boolean = if (exception.isInstanceOf[TransientException]) {
```
should probably a pattern match in the future

conker84 replied:
Added in this one

jexp commented on src/main/scala/org/neo4j/spark/utils/Neo4jUtils.scala:
```diff
@@ -0,0 +1,58 @@
+    }
+  }
+
+  private val retryConfig = RetryConfig.custom.retryExceptions(
```
wouldn't that be usually be something that's handled by Spark? How do others handle this? Just curious.

conker84 replied:
The spark retry mechanism works with `spark.task.maxFailures` which is:
```
Number of failures of any particular task before giving up on the job.
The total number of failures spread across different tasks will not cause the job to fail;
a particular task has to fail this number of attempts
```

But IMHO the control it's not fine-graded as this one, I would like to have both of them, wdyt?

jexp replied:
perhaps make it configurable? (later PR)

conker84 commented:
@jexp I updated the PR please check my answers to the comments.
Thanks a lot for the feedback!!!

jexp commented on src/main/scala/org/neo4j/spark/utils/Neo4jUtils.scala:
```diff
@@ -0,0 +1,71 @@
+  def close(driver: Driver, session: Session): Unit = {
+    try {
+      if (session != null && session.isOpen) {
+        session.close()
```
Please add the exception handling as we discussed.

conker84 replied:
done

jexp commented on src/test/scala/org/neo4j/spark/Neo4jSparkTSE.scala:
```diff
@@ -85,18 +74,18 @@ class Neo4jSparkTest {
 
   @Test def runCypherQueryWithPartition() {
-    val neo4j: Neo4j = Neo4j(sc).cypher("MATCH (n:Person) RETURN id(n) SKIP {_skip} LIMIT {_limit}").partitions(4).batch(25)
+    val neo4j: Neo4j = Neo4j(sc).cypher("MATCH (n:Person) RETURN id(n) SKIP $_skip LIMIT $_limit").partitions(4).batch(25)
```
did we also change the test inside of the execution/partition?
and the docs in the readme?

conker84 replied:
yes in both of them

jexp commented on src/test/scala/org/neo4j/spark/SparkConnectorScalaBaseTSE.scala:
```diff
@@ -0,0 +1,64 @@
+  def before() {
+    SparkConnectorScalaSuiteIT.session()
+      .writeTransaction(new TransactionWork[ResultSummary] {
+        override def execute(tx: Transaction): ResultSummary = tx.run("MATCH (n) DETACH DELETE n").consume()
```
does this work with larger test datasets?

conker84 replied:
no, we must use a different strategy

jexp commented on README.md:
```diff
@@ -14,29 +14,29 @@ This neo4j-spark-connector is Apache 2 Licensed
 
 
-Build `target/neo4j-spark-connector_2.11-full-2.4.1-M1.jar` for Scala 2.11
+Build `target/neo4j-spark-connector_2.11-full-2.6.0-M1.jar` for Scala 2.11
```
2.4.0-M2

jexp replied:
2.4.4-M2

jexp commented on pom.xml:
```diff
@@ -4,7 +4,7 @@
     <groupId>neo4j-contrib</groupId>
     <packaging>jar</packaging>
-    <version>2.4.1-M1</version>
+    <version>2.6.0-M1</version>
```
2.4.4-M2

jexp commented on src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala:
```diff
@@ -287,12 +212,11 @@ object CypherTypes {
-    else if (typ == typeSystem.TIME()) CypherTypes.TIME
-    else if (typ == typeSystem.LOCAL_TIME()) CypherTypes.TIME	
+    else if (typ == typeSystem.TIME()) CypherTypes.DATETIME
+    else if (typ == typeSystem.FLOAT()) CypherTypes.FlOAT
```
where did typeSystem.LOCAL_TIME go?

jexp commented on src/main/scala/org/neo4j/spark/utils/Neo4jSessionAwareIterator.scala:
```diff
@@ -0,0 +1,70 @@
+          transaction.commit()
+        }
+      }
+      if (result != null) {
```
are you sure that this should happen after tx.commit / rollback? I'm not sure, perhaps ask Nigel?

jexp commented on src/test/scala/org/neo4j/spark/Neo4jDataFrameScalaTSE.scala:
```diff
@@ -44,9 +28,8 @@ class Neo4jDataFrameScalaTest {
-    val it: ResourceIterator[Long] = server.graph().execute("MATCH (:Person {name:'Keanu'})-[:ACTED_IN]->(:Movie {title:'Matrix'}) RETURN count(*) as c").columnAs("c")
-    assertEquals(1L, it.next())
-    it.close()
+    val count = SparkConnectorScalaSuiteIT.session().run("MATCH (:Person {name:'Keanu'})-[:ACTED_IN]->(:Movie {title:'Matrix'}) RETURN count(*) as c").single().get("c").asLong()
```
consider using triple-quotes for these long statements, but that can also happen in a later PR
